File: frontend/src/lib/api/exercises.ts
import { api } from "./http";
import type {
  Exercise,
  ExerciseFilters,
  ExerciseInput,
  ExerciseQuery,
} from "@/types/exercise";

function buildQueryString(query: ExerciseQuery): string {
  const params = new URLSearchParams();

  for (const [key, value] of Object.entries(query)) {
    if (value && value.trim() !== "") {
      params.set(key, value.trim());
    }
  }

  const queryString = params.toString();
  return queryString === "" ? "" : `?${queryString}`;
}

/** The backend takes the scalar fields as form values and JSON-encoded string lists. */
function buildFormData(input: Partial<ExerciseInput>): FormData {
  const formData = new FormData();

  const appendIfPresent = (key: string, value: string | null | undefined) => {
    if (value !== undefined && value !== null && value !== "") {
      formData.append(key, value);
    }
  };

  appendIfPresent("name_es", input.name_es);
  appendIfPresent("category_es", input.category_es);
  appendIfPresent("level_es", input.level_es);
  appendIfPresent("force_es", input.force_es);
  appendIfPresent("mechanic_es", input.mechanic_es);
  appendIfPresent("equipment_es", input.equipment_es);

  if (input.instructions_es) {
    formData.append("instructions_es", JSON.stringify(input.instructions_es));
  }
  if (input.primary_muscles_es) {
    formData.append(
      "primary_muscles_es",
      JSON.stringify(input.primary_muscles_es),
    );
  }
  if (input.secondary_muscles_es) {
    formData.append(
      "secondary_muscles_es",
      JSON.stringify(input.secondary_muscles_es),
    );
  }

  for (const image of input.images ?? []) {
    formData.append("images", image);
  }

  return formData;
}

export function listExercises(query: ExerciseQuery = {}): Promise<Exercise[]> {
  return api.get<Exercise[]>(`/exercises${buildQueryString(query)}`);
}

export function getExercise(exerciseId: string): Promise<Exercise> {
  return api.get<Exercise>(`/exercises/${exerciseId}`);
}

export function getExerciseFilters(): Promise<ExerciseFilters> {
  return api.get<ExerciseFilters>("/exercises/filters");
}

export function createExercise(input: ExerciseInput): Promise<Exercise> {
  return api.postForm<Exercise>("/exercises", buildFormData(input));
}

export function updateExercise(
  exerciseId: string,
  input: Partial<ExerciseInput>,
): Promise<Exercise> {
  return api.patchForm<Exercise>(
    `/exercises/${exerciseId}`,
    buildFormData(input),
  );
}

export function deleteExercise(exerciseId: string): Promise<void> {
  return api.delete(`/exercises/${exerciseId}`);
}

File: backend/app/api/exercises.py
from fastapi import APIRouter, Depends, File, Form, UploadFile, status
from sqlalchemy.orm import Session

from app.core.db import get_db
from app.core.security import get_current_trainer
from app.models import Exercise, Trainer
from app.schemas.exercise import ExerciseFiltersOut, ExerciseOut
from app.services import exercise_service

router = APIRouter(
    prefix="/api/exercises",
    tags=["exercises"],
    dependencies=[Depends(get_current_trainer)],
)


@router.get("", response_model=list[ExerciseOut])
def list_exercises(
    muscle: str | None = None,
    equipment: str | None = None,
    category: str | None = None,
    level: str | None = None,
    search: str | None = None,
    db: Session = Depends(get_db),
) -> list[Exercise]:
    return exercise_service.list_exercises(
        db,
        muscle=muscle,
        equipment=equipment,
        category=category,
        level=level,
        search=search,
    )


@router.get("/filters", response_model=ExerciseFiltersOut)
def get_filters(db: Session = Depends(get_db)) -> dict[str, list[str]]:
    return exercise_service.get_filters(db)


@router.get("/{exercise_id}", response_model=ExerciseOut)
def get_exercise(exercise_id: str, db: Session = Depends(get_db)) -> Exercise:
    return exercise_service.get_exercise(db, exercise_id)


@router.post("", response_model=ExerciseOut, status_code=status.HTTP_201_CREATED)
async def create_exercise(
    name_es: str = Form(...),
    instructions_es: str = Form(..., description="JSON array of strings"),
    category_es: str | None = Form(None),
    level_es: str | None = Form(None),
    force_es: str | None = Form(None),
    mechanic_es: str | None = Form(None),
    equipment_es: str | None = Form(None),
    primary_muscles_es: str | None = Form(None, description="JSON array of strings"),
    secondary_muscles_es: str | None = Form(None, description="JSON array of strings"),
    images: list[UploadFile] = File(default=[]),
    trainer: Trainer = Depends(get_current_trainer),
    db: Session = Depends(get_db),
) -> Exercise:
    return await exercise_service.create_exercise(
        db,
        trainer,
        name_es=name_es,
        instructions_es=exercise_service.parse_json_string_list(
            instructions_es, "instructions_es"
        ),
        category_es=category_es,
        level_es=level_es,
        force_es=force_es,
        mechanic_es=mechanic_es,
        equipment_es=equipment_es,
        primary_muscles_es=exercise_service.parse_json_string_list(
            primary_muscles_es, "primary_muscles_es"
        ),
        secondary_muscles_es=exercise_service.parse_json_string_list(
            secondary_muscles_es, "secondary_muscles_es"
        ),
        images=images,
    )


@router.patch("/{exercise_id}", response_model=ExerciseOut)
async def update_exercise(
    exercise_id: str,
    name_es: str | None = Form(None),
    instructions_es: str | None = Form(None, description="JSON array of strings"),
    category_es: str | None = Form(None),
    level_es: str | None = Form(None),
    force_es: str | None = Form(None),
    mechanic_es: str | None = Form(None),
    equipment_es: str | None = Form(None),
    primary_muscles_es: str | None = Form(None, description="JSON array of strings"),
    secondary_muscles_es: str | None = Form(None, description="JSON array of strings"),
    images: list[UploadFile] | None = File(default=None),
    trainer: Trainer = Depends(get_current_trainer),
    db: Session = Depends(get_db),
) -> Exercise:
    updates: dict = {}
    if name_es is not None:
        updates["name_es"] = name_es
    if instructions_es is not None:
        updates["instructions_es"] = exercise_service.parse_json_string_list(
            instructions_es, "instructions_es"
        )
    if category_es is not None:
        updates["category_es"] = category_es
    if level_es is not None:
        updates["level_es"] = level_es
    if force_es is not None:
        updates["force_es"] = force_es
    if mechanic_es is not None:
        updates["mechanic_es"] = mechanic_es
    if equipment_es is not None:
        updates["equipment_es"] = equipment_es
    if primary_muscles_es is not None:
        updates["primary_muscles_es"] = exercise_service.parse_json_string_list(
            primary_muscles_es, "primary_muscles_es"
        )
    if secondary_muscles_es is not None:
        updates["secondary_muscles_es"] = exercise_service.parse_json_string_list(
            secondary_muscles_es, "secondary_muscles_es"
        )

    return await exercise_service.update_exercise(
        db, trainer, exercise_id, updates=updates, images=images
    )


@router.delete("/{exercise_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_exercise(
    exercise_id: str,
    trainer: Trainer = Depends(get_current_trainer),
    db: Session = Depends(get_db),
) -> None:
    exercise_service.delete_exercise(db, trainer, exercise_id)
